platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ExportMigrationContextImpl.java

```diff
@@ -0,0 +1,278 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.configuration.migration;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.function.BiPredicate;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.filefilter.TrueFileFilter;
+import org.apache.commons.io.output.ClosedOutputStream;
+import org.apache.commons.io.output.ProxyOutputStream;
+import org.apache.commons.lang.Validate;
+import org.codice.ddf.migration.ExportMigrationContext;
+import org.codice.ddf.migration.ExportMigrationEntry;
+import org.codice.ddf.migration.Migratable;
+import org.codice.ddf.migration.MigrationException;
+import org.codice.ddf.migration.MigrationReport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableMap;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * The export migration context keeps track of exported migration entries for a given migratable
+ * while processing an export migration operation.
+ */
+public class ExportMigrationContextImpl extends MigrationContextImpl
+        implements ExportMigrationContext, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportMigrationContextImpl.class);
+
+    /**
+     * Holds exported migration entries keyed by the exported path.
+     */
+    private final Map<Path, ExportMigrationEntryImpl> entries = new TreeMap<>();
+
+    /**
+     * Holds migration entries referenced from system properties keyed by the property name.
+     */
+    private final Map<String, ExportMigrationSystemPropertyReferencedEntryImpl> systemProperties =
```

Comment:
❓ Could we use `ExportMigrationEntry` in the `Map` (and everywhere else we're using it right now) instead of the concrete class `ExportMigrationSystemPropertyReferencedEntryImpl`?

Reply:
this one I could, some others I could not without type casting it everywhere.